server/grpc_service.go

```diff
@@ -908,3 +908,106 @@ func (s *Server) incompatibleVersion(tag string) *pdpb.ResponseHeader {
 		Message: msg,
 	})
 }
+
+// PrewriteMaxTS is the first part of 2PC for global MaxTS synchronization,
+// it compares the MaxTS received with each Local TSO Allocator leader's TSO
+// and prewrite it into memory to wait for the next phase of formal writing.
+func (s *Server) PrewriteMaxTS(ctx context.Context, request *pdpb.PrewriteMaxTSRequest) (*pdpb.PrewriteMaxTSResponse, error) {
+	if err := s.validateInternalRequest(request.GetHeader()); err != nil {
+		return nil, err
+	}
+	tsoAllocatorManager := s.GetTSOAllocatorManager()
+	// Get all Local TSO Allocator leaders
+	allocatorLeaders, err := tsoAllocatorManager.GetLocalAllocatorLeaders()
+	if err != nil {
+		return nil, err
+	}
+	// Iterate all Local TSO Allocator leaders to get the max local TSO
+	var maxLocalTS pdpb.Timestamp
+	for _, allocator := range allocatorLeaders {
+		currentLocalTSO, err := allocator.GetCurrentTSO()
+		if err != nil {
+			return nil, err
+		}
+		if currentLocalTSO.Physical > maxLocalTS.Physical {
+			maxLocalTS.Physical = currentLocalTSO.Physical
+		}
+	}
+	// Validate whether the MaxTs is greater than all other local TSOs
+	if maxLocalTS.Physical > request.MaxTs.Physical {
+		return &pdpb.PrewriteMaxTSResponse{
+			Header:     s.header(),
+			Prewritten: false,
+			MaxLocalTs: &maxLocalTS,
+		}, nil
+	}
+	// Do the prewriting
+	for _, allocator := range allocatorLeaders {
+		allocator.PrewriteTSO(request.MaxTs)
+	}
+	return &pdpb.PrewriteMaxTSResponse{
+		Header:     s.header(),
+		Prewritten: true,
+	}, nil
+}
+
+// WriteMaxTS is the second part of 2PC for global MaxTS synchronization,
+// it compares the MaxTS received with each Local TSO Allocator leader's
+// prewritten TSO in memory and write it into memory to finish the global
+// TSO synchronization progress.
+func (s *Server) WriteMaxTS(ctx context.Context, request *pdpb.WriteMaxTSRequest) (*pdpb.WriteMaxTSResponse, error) {
+	if err := s.validateInternalRequest(request.GetHeader()); err != nil {
+		return nil, err
+	}
+	tsoAllocatorManager := s.GetTSOAllocatorManager()
+	// Get all Local TSO Allocator leaders
+	allocatorLeaders, err := tsoAllocatorManager.GetLocalAllocatorLeaders()
+	if err != nil {
+		return nil, err
+	}
+	// Iterate all Local TSO Allocator leaders to validate whether the prewritten TSO is consistent
```

Comment:
Why do we need the prewrite tso? seems we can use the ts in `request` directly?

Reply:
What if it's a request that arrives very late that the prewritten TSO is no longer valid during an abortive transaction?

Reply:
The prewritten TSO is a bridge connecting the previous Prewrite request and the later Write request. Otherwise the requests will be separated, which make a Local TSO Allocator leader not know whether it's a correct request to write the corresponding prewritten TSO.